File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    //could just fully brute force it with like checking
    //each character agaisnt each other and then keeping a count when the next character is not like the character before

    //need to create some kind of set or map
    const seen = new Map();

    //kep track of the starting index for the current substring
    let start = 0;
    
    //keep track of max length
    let longest = 0;

    for(let i = 0; i < s.length; i++){
        //have we seen the current char
        //if we have move the start 1 + the last index of this char
        if(seen.has(s[i])){
            start = Math.max(seen.get(s[i]) + 1, start);
        }
        seen.set(s[i], i);
        
        //longest = current sub length and the max length
        longest = Math.max(i - start + 1, longest);
    }


    return longest;


};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        //use 2 pointers
        //start both pointers at the start of the array and 

        //use a hash set this will keep track of the seen values and make sure theres only one of it

        //useing a set is ok here casue there are only 26 possible elements which means sets can work ok even if it means constant checking of the set

        int maxLen = 0;

        HashSet<Character> set = new HashSet<>();

        int left = 0;

        for(int right = 0; right<s.length(); right++){
            //look at the current character in the list and check if we have seen it or not
            char current = s.charAt(right);

            //check if the set contains the current character
            //will not run if the set doesn't have the character in it
            while(set.contains(current)){
                //if the character is in the set remove it and move the left pointer over as it means there is a dupe
                set.remove(s.charAt(left));
                left++;

            }

            //if the letter is not in the list add it to the list
            set.add(current);

            //get the current max substring length
            maxLen = Math.max(maxLen,right-left+1); //+1 cause left starts at 0

        }

        return maxLen;

        

    }
}